Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -479,4 +479,170 @@ HRESULT _DWriteFontGetBoundingBoxesForGlyphs(
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetFileSize(_Out_ uint64_t* fileSize) {
+        RETURN_HR_IF_NULL(E_INVALIDARG, fileSize);
+        *fileSize = CFDataGetLength(_data.get());
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE GetLastWriteTime(_Out_ uint64_t* lastWriteTime) {
+        RETURN_HR_IF_NULL(E_INVALIDARG, lastWriteTime);
+        *lastWriteTime = _lastWriteTime;
+        return S_OK;
+    };
+
+    HRESULT STDMETHODCALLTYPE ReadFileFragment(_Out_ const void** fragmentStart,
+                                               uint64_t fileOffset,
+                                               uint64_t fragmentSize,
+                                               _Out_ void** fragmentContext) {
+        if (fileOffset + fragmentSize > CFDataGetLength(_data.get())) {
+            return E_INVALIDARG;
+        }
+
+        if (fragmentStart) {
+            const uint8_t* underlyingBuffer = CFDataGetBytePtr(_data.get());
+            *fragmentStart = reinterpret_cast<const void*>(&(underlyingBuffer[fileOffset]));
+        }
+
+        if (fragmentContext) {
+            // Deliberately unused: this is meant to be passed to ReleaseFileFragment() below to free part of the data,
+            // but this stream frees all the underlying CFData at once
+            *fragmentContext = nullptr;
+        }
+
+        return S_OK;
+    };
+
+    void STDMETHODCALLTYPE ReleaseFileFragment(void* fragmentContext){
+        // Deliberate no-op: data is released alongside the _data member with the destruction of this object
+    };
+
+private:
+    woc::unique_cf<CFDataRef> _data;
+    uint64_t _lastWriteTime;
+};
+
+/**
+ * Custom implementation of IDWriteFontFileLoader that loads a CGDataProviderRef as its font file
+ */
+class DWriteFontBinaryDataLoader : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontFileLoader> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontBinaryDataLoader() {
+    }
+
+    HRESULT RuntimeClassInitialize(CGDataProviderRef dataProvider) {
+        CFRetain(dataProvider);
+        _dataProvider.reset(dataProvider);
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReferenceKey,
+                                                  uint32_t fontFileReferenceKeySize,
+                                                  _Out_ IDWriteFontFileStream** fontFileStream) {
+        if (!fontFileStream) {
+            return E_INVALIDARG;
+        }
+
+        // Ignore first two params, just return the same kind of stream always
+        ComPtr<DWriteFontBinaryDataStream> ret;
+        RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataStream>(&ret, _dataProvider.get()));
+
+        *fontFileStream = ret.Detach();
+        return S_OK;
+    }
+
+private:
+    woc::unique_cf<CGDataProviderRef> _dataProvider;
+};
```

Comment:
DWriteWrapper_CoreText is also over 900 lines, and has two classes like this.

I don't want to move just this into a new file right now, and ignore the other two, because that fractures our codebase along conflicting principles.
I also don't want to move all four classes into (a? two? four?) new files right now, because three people have already signed off on this PR, and it'd be easier to get them to look at/provide feedback on large-scale reorganizational changes in a new PR.
This pull request is also rapidly scope-creeping for a change that was supposed to implement one (!) function that was regressed during CoreText bringup.

As a general principle, I prefer more & smaller PRs, to fewer & larger ones. Large changes make it much more difficult to pinpoint what exactly went wrong, when something does. I understand that some things won't be done if they're not attached to larger pull requests. However, this is a case where I have an explicit issue for this. I feel extremely strongly that I would rather move the four classes to new classes (assuming that's what other people want!) in a new PR, rather than try to get them to look once more at a PR that they've already approved.

---
In reply to: [88313117](https://github.com/Microsoft/WinObjC/pull/1386#discussion_r88313117) [](ancestors = 88313117,88312848,88280909,88148206)